src/RazorLight/Compilation/RoslynCompilationService.cs

```diff
@@ -22,11 +22,13 @@ public class RoslynCompilationService : ICompilationService
 		private readonly IMetadataReferenceManager metadataReferenceManager;
 		private readonly bool isDevelopment;
 		private readonly List<MetadataReference> metadataReferences = new List<MetadataReference>();
+		private readonly List<Action<IGeneratedRazorTemplate, byte[], byte[]>> precreateAssemblyCallbacks;
 
-		public RoslynCompilationService(IMetadataReferenceManager referenceManager, Assembly operatingAssembly)
+		public RoslynCompilationService(IMetadataReferenceManager referenceManager, Assembly operatingAssembly, List<Action<IGeneratedRazorTemplate, byte[], byte[]>> precreateAssemblyCallbacks = null)
```

Comment:
1. The interface here is really not "discoverable" - I have worked on this code base for over 4 years now, and just looking at this callback makes my brain hurt. The lack of a default implementation also further increases the odds people will use it incorrectly or copy-paste some bad code from somewhere on the Internet, and we will get a lot of support Issues dealing with lazy programming. - This is **the top** maintenance cost of this project thus far.
2. Why are you calling it precreate vs. precompile?
3. If your goal is to precompile assemblies, why not contribute a PR to the RazorLight.Precompile csproj?

Reply:
My reply would be this:

1. How is it different from the prerender callbacks?
2. My rationale is that it is called before the Assembly is created. The code has already been compiled, so it does not happen before the compilation. But if `precompile` is more recognizable, then it makes sense to rename. I guess this has to do with discoverability. `precompile` is a known buzz word, `precreate` is not. So, makes sense to rename. Unless, we want to keep it obscured :-).
3. I do not know. I was not aware of this tool. I did notice the project, but since it was not mentioned in the documentation (or I missed it) I just ignored it. What is the state of this tool?
4. Makes sense in general. 

TBH, I am acting on my selfish reasons - I need to implement something in Powershell yesterday and as it is either I drop RazorLight entirely (in favour of what exactly? as far as I know there is nothing being as popular and actively maintained) or complicate a straightforward and conceptually simple scripting code to move all the rendering to one executable call.

If `RazorLight.Precompile` is already usable, then definitely I would prefer using it. But it may be a longer engagement and in the meantime - the ability to cache generated assemblies is needed now. At least I need it now.

Can we find a compromise that would allow us to release something quickly yet satisfying the quality requirements? I will gladly implement all your recommendations, if at the end I will be able to achieve my goal in a timely fashion.

Hope it makes sense.

Reply:
> If `RazorLight.Precompile` is already usable, then definitely I would prefer using it. But it may be a longer engagement and in the meantime - the ability to cache generated assemblies is needed now. At least I need it now.

I'm not trying to create work for you, but it seems straight forward to lift your PR into a tool vs. something that caches everything on startup.  The RazorLight.Precompile project is just a stub right now.

> if at the end I will be able to achieve my goal in a timely fashion.

I'm not trying to create work, but I do think most people would rather precompile this stuff vs. create a persistent cache. It makes the entire flow a lot more debuggable since you have to use the compiled DLLs on disk vs. use a persistent cache that could get turned off via misconfiguration or load a different projects cached items by accident.